modules/admin_manual/pages/configuration/server/security/hsmdaemon.adoc

```diff
@@ -0,0 +1,588 @@
+= The HSM Daemon (hsmdaemon)
+:base64-encoding-url: https://en.wikipedia.org/wiki/Base64
+:hsm-url: https://en.wikipedia.org/wiki/Hardware_security_module
+:jwt-url: https://jwt.io/
+:network-sockets-url: https://en.wikipedia.org/wiki/Network_socket
+:go-install-url: https://golang.org/doc/install
+:opensc-wiki-url: https://github.com/OpenSC/OpenSC/wiki
+:opensuse-security-repositories-url: https://download.opensuse.org/repositories/security/
+:php-exec-function-url: https://www.php.net/manual/en/function.exec.php
+:pkcs11-url: https://en.wikipedia.org/wiki/PKCS_11
+:pkcs11-tool-url: https://linux.die.net/man/1/pkcs11-tool 
+:softhsm2-url: https://www.opendnssec.org/softhsm/
+:unix-sockets-url: http://beej.us/guide/bgipc/html/multi/unixsock.html
+
+The hsmdaemon is a daemon to delegate encryption to an {hsm-url}[HSM (Hardware Security Module)].
+It can be necessary, as PHP cannot directly interface with {pkcs11-url}[a PKCS11 stack]; neither by an API wrapper, because one does not exist, yet, nor via the OpenSSL bindings
+Because of this, a separate process is needed to decrypt anything with the private key stored in an HSM. 
+
+Running {php-exec-function-url}[exec()] to decrypt the key with a command line command to do the encryption might leak the HSM credentials if the admin lists the currently running processes. 
+To prevent that an "HSM" daemon will be used that can open a session to the HSM upon startup. 
+
+This daemon will be used by ownCloud to decrypt the current master key upon request. 
+The communication happens via {unix-sockets-url}[UNIX sockets] or {network-sockets-url}[TCP sockets] and is authorized by a shared token that the daemon stores in the ownCloud database via a REST/JSON route.
+
+ownCloud internally uses OpenSSL to en-/decrypt keys and needs to be extended to support en-/decrypt operations via the new daemon. 
+The current solution would be to encrypt the current ownCloud master key with a key from the HSM. 
```

Comment:
Do I read correct, that you encrypt a key by a key or is it meant the the keys are replaced

Reply:
Not sure. Checking.

Reply:
@butonic, can you answer this please?

Reply:
Not exactly. The keys are not replaced. From the technical point of view the Crypt class is extended to handle the key generation in the HSM device and also to get the key from HSM. So basically for the read/write operation on a file, the request goes to the HSM and then based on the keys fetched from HSM the files are encrypted/decrypted. Does that answer your query?

Reply:
Thanks for the further input, @sharidas.

Reply:
If fe414b4efa8ce8cb718ec648184bf55b2d64418c is okay, @sharidas, can you approve the PR, pls?

Reply:
Looks ok to me :+1:

Reply:
Sorry I need to insist...
The answer to my question is not clear nor the impact.
It currently says that you encrypt oC masterkey with a key from HSM.
Is it what it reads or is it different. What does this mean. What is the impact of a key re-encrypted by HSM.
The sentence also says „would be“. This hands over the insecurity of the process to the admin and also tells that we do not know. Either we say it clear - this is the process (with a description of it) or we rewrite this statement and describe the opportunities/possibilities and their impact.

Reply:
@sharidas, can you clarify, pls?

Reply:
@mmattel, can you have another look and let me know if the change resolves your concerns?

Reply:
Concern solved, thanks for taking care of. Marking as resolved